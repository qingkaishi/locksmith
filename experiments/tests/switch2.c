#include <pthread.h>
#include <stdio.h>

int *j;
pthread_mutex_t l;

int f() {
  *j = 3;
}

int (*ptr)(void);

void * dostuff(void * a) {
  int k;
  
  pthread_mutex_lock(&l);
  j = &k;
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1;
  pthread_mutex_init(&l, NULL);
  int i = 1;

  ptr = f;
  pthread_create(&t1, NULL, dostuff, NULL);
//  pthread_create(&t1, NULL, dostuff, NULL);
  switch(i) {
    case 1:
      break;
    default:
      i = 2;
      ptr(); //a race
      break;
      j=&i; //this should not be in the guard set
  }
}


