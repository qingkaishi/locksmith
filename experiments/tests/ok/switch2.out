************************* STARTING *************************

shared: &j:switch2.c:4

Warning: Possible data race: &j:switch2.c:4 is not protected!
 references:
  dereference of &j:switch2.c:4 at switch2.c:8
    &j:switch2.c:4
  locks acquired:
    <empty>
  in: main at switch2.c:22  -> switch2.c:35

  dereference of &j:switch2.c:4 at switch2.c:17
    &j:switch2.c:4
  locks acquired:
    concrete l:switch2.c:24
    l:switch2.c:5
  in: main at switch2.c:22 


*************************** DONE ***************************
